Add very basic unit conversion

SimonHeybrock opened the pull request:
This adds the first example of a time-of-flight unit conversion. This demonstrates the overall structure of the actual conversion math, but many details are incomplete.

If you compare this to how Mantid unit conversion works:
- We are computing the conversion factors for the whole array (all spectra) at once.
- We apply the conversion factor using our normal binary operations, which handle broadcasting, etc.

Many of the TODO items depend on other open PR or tasks that have not been started yet, so they will have to wait.

We will also need to extract the common conversion code and provide mechanism to dispatch to the correct conversion implementation.

SimonHeybrock commented on src/dataset.cpp:
```diff
@@ -514,8 +515,81 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+Dataset tofToEnergy(const Dataset &d) {
+  // 1. Compute conversion factor
+  const auto &compPos =
+      d.get<const Coord::ComponentInfo>()[0].get<const Coord::Position>();
```
`get<>` will throw if it does not find the tag, i.e., includes the check. Maybe we would want to catch that and give a more specific error message, but that should be made consistent across all other API functions as well.

SimonHeybrock commented on src/dataset.cpp:
```diff
@@ -514,8 +515,81 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+
 Dataset convert(const Dataset &d, const Dim from, const Dim to) {
-  static_cast<void>(to);
+  if ((from == Dim::Tof) && (to == Dim::Energy))
```
I still have to figure out how to do this. Mantid currently always converts via TOF (unless there are special cases enabled), i.e., when you request conversion from A to B it does A -> TOF -> B. Ideally I would like to avoid this, but it is not clear to me yet whether we can do that with help of the compiler, or whether we would have to hand-code all possible conversions.

The specialization of `convert`, or how it dispatches to the actual conversion method will depend on that... so I think we have to postpone this until the problem is better understood and refactor then.

SimonHeybrock commented on src/except.cpp:
```diff
@@ -55,6 +57,8 @@ std::string to_string(const Tag tag) {
   switch (tag.value()) {
   case Coord::Tof{}.value():
     return "Coord::Tof";
+  case Coord::Energy{}.value():
```
A tag is a very small object (currently 2 Byte). Depending on how we will end up using `Variable` and tags, I would prefer to keep it that way, not adding the 8 Byte (effectively 16 Byte) overhead to `Variable`, just for having the pointer to the vtable.

OwenArnold replied:
Sounds like something that Andrei Alexandrescu might have a solution for, type_list etc.

SimonHeybrock replied:
Tags are likely to be refactored anyway, see https://github.com/mantidproject/dataset/blob/master/doc/design-tag-considerations.md.

OwenArnold commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+      // The reshape is just to remap the dimension label, should probably do
+      // this differently.
+      energy /= (var * var).reshape(varDims);
+      converted.insert(energy);
```
If i'm reading this correctly the approach is to transform any variable matching the criteria into a new energy variable for output. Did you consider offering this as an in-place operation?

SimonHeybrock replied:
This `if` branch is only for the coordinate. Typically that data is small compared to the other data. We thus should first solve to issue discussed in a code comment below regarding changing dimension labels without trigger copies.

Regarding in-place operations: What I do currently in several operations is to take the `Dataset` argument by value, so in C++ we can `std::move` the input we we want in-place operation. On the Python side this is not supported currently.

OwenArnold commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+    auto varDims = var.dimensions();
+    if (varDims.contains(Dim::Tof))
+      varDims.relabel(varDims.index(Dim::Tof), Dim::Energy);
+    if (var.tag() == Coord::Tof{}) {
```
Are you following the approach currently taken by Mantid of having a "base" unit i.e. all units are converted to-from TOF? If not, then direct wavelength->energy conversions would be possible, and faster.

OwenArnold commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+    } else {
+      // TODO Changing Dim::Tof to Dim::Energy. Currently this is not possible
+      // without making a copy of the data, which is inefficient. If we cannot
+      // move the dimensions outside the cow_ptr in Variable, another option
```
Ok, this relates to my comment above.

Presumably you would support in-place operations via an overload for each of the converters. `void tofToEnergy(Dataset &d)`

OwenArnold replied:
Also, can we move these converter free functions out of `dataset` maybe something dedicated to unit conversion.

SimonHeybrock replied:
Eventually the unit conversion is likely to be moved to a separate library.

OwenArnold commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+    tofShift.setDimensions(d(Coord::Ei{}).dimensions());
+    const auto &Ei = d.get<const Coord::Ei>();
+    std::transform(Ei.begin(), Ei.end(), tofShift.get<Data::Value>().begin(),
+                   [&](const double Ei) { return l1 / sqrt(Ei); });
```
minor: [&] -> [&l1]

OwenArnold commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+      "Conversion between requested dimensions not implemented yet.");
   // How to convert? There are several cases:
   // 1. Tof conversion as Mantid's ConvertUnits.
   // 2. Axis conversion as Mantid's ConvertSpectrumAxis.
```
I'm not sure we need to make a distinction. The whole vertical axis thing seems tied up with the Workspace2D structure. Isn't this now just a case of what conversions are defined possible. i.e `spectrum_position` -> `two_theta`, `tof->wavelength`

SimonHeybrock replied:
Exactly, that was also my idea. But for the spectrum axis there is a slightly complication: It essentially starts off as a discrete, non-continuous axis, but can be converted into one.

It is also not so clear of for `ConvertSpectrumAxis` we actually *want* to convert the dimension: In the `Dataset` way of operation, all data would then not have `Dim::Spectrum` but `Dim::TwoTheta` (or similar). It thus cannot be indexed by spectrum number any more, etc., ... The alternative would be to simply add a new coordinate `Coord::TwoTheta` but with `Dim::Spectrum` (*not* `Dim::TwoTheta`).

SimonHeybrock replied:
On the other hand, maybe we should have something like `Dim::Position` as the common "primary" dimension, instead of `Dim::Spectrum`? The latter is after all just the dimension we use as a matter of habit, not because it is the natural representation of data? We could of course still have `Coord::SpectrumNumber`, but it would have `Dim::Position`.

Would this description work also well for other data where spectra are not linked to positions? For example, after `SofQw` we would have `Dim::Q` (instead of `Dim::Spectrum`). The disadvantage I can see is that data now has a different dimension label depending on what it represents. But maybe that is actually an advantage? After all I am proposing to do the same for `Dim::Tof` which would be something else after unit conversion.

nvaytet replied:
I agree that a `Dim::Position` makes more sense to me than `Dim::Spectrum` which is confusing to anyone new to Mantid. It also makes it more general, if `Dataset` is ever to be applied to something other than neutrons or Mantid. But then you also have `Coord::Position`, so this might make things more confusing...

SimonHeybrock replied:
`Dim::Position` and `Coord::Position` are linked.

Regarding confusing people new to Mantid: I do not know if `Dim::Position` is less confusing, you can ask "position of what"?

OwenArnold commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+  }
+
+  return converted;
+}
```
Given the flexibility of what can be done here, in this units like approach the implementation of TOF -> Q3D (histogram) would look very similar.  Potentially, `Q1D`, `Qxy`, `SofQW`, `ConvertToReflectometryQ` .... could also be done in this form.

SimonHeybrock replied:
Yes, see the long comment in the body of the `convert` function. The main question is whether the conversion is uniquely specified by providing dimensions?

For example:

```cpp
auto q1d = convert(dataset, {Dim::Tof, Dim::Spectrum}, {Dim::Q});
auto qxy = convert(dataset, {Dim::Tof, Dim::Spectrum}, {Dim::Qx, Dim::Qy});
```

Is this completely unambiguous?

OwenArnold replied:
Given your `Ei`, `Ef` improvements, I think it's definitely easier to deduce what is possible, though for cases such as elastic conversions, you may also (optionally) convert into different frames (QLab, QSample, etc).

SimonHeybrock replied:
Would that just be a different dimension, `Dim::QLabel` and `Dim::QSample`?

OwenArnold commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+}
+} // namespace tof
+} // namespace neutron
+
```
I don't see why the `from` is strictly necessary.

SimonHeybrock replied:
To specify which dimension should be converted.

SimonHeybrock commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+  Variable conversionFactor(Data::Value{}, dims);
+
+  MDZipView<const Coord::Position> specPos(d);
+  const auto factor = [&](const auto &item) {
```
It also contain physical constants (or rather, will contain).

nvaytet commented on src/dataset.cpp:
```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
   return out;
 }
 
+namespace neutron {
```
Just a quick comment, what do you think about moving unit conversions to a seperate file, instead of the main `dataset.cpp` implementation? We might end up having quite a few comversions in the end, it feels like it's kind of polluting `dataset.cpp` a little...

SimonHeybrock commented:
I think I have addressed the comments that can be at this point. Others such as unit scales depend on a proper implementation of units, so it will have to wait. Later on there will definitely be a couple of follow-up PRs.
